transform-plugins/src/main/java/co/cask/hydrator/plugin/ValueMapper.java

```diff
@@ -0,0 +1,251 @@
+      Lookup<String> lookupTable = context.provide(mapping.getLookupTableName(), arguments.getProperties());
+      mapping.setLookupTable(lookupTable);
+    }
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+
+    StructuredRecord.Builder builder = StructuredRecord.builder(getOutputSchema(input.getSchema()));
+
+    for (Schema.Field sourceField : input.getSchema().getFields()) {
+      if (mappingValues.containsKey(sourceField.getName())) {
+        ValueMapping mapping = mappingValues.get(sourceField.getName());
+        if (input.get(sourceField.getName()) == null || input.get(sourceField.getName()).toString().isEmpty()) {
+          if (mapping.getDefaultValue() != null) {
+            builder.set(mapping.getTargetField(), mapping.getDefaultValue());
+          } else {
+            builder.set(mapping.getTargetField(), input.get(sourceField.getName()));
+          }
+        } else {
+          // for those source field whose values are neither NULL nor EMPTY
+          Lookup<String> valueMapperLookUp = mapping.getLookupTable();
+          String lookupValue = valueMapperLookUp.lookup(input.get(sourceField.getName()).toString());
+          if (lookupValue != null && !lookupValue.isEmpty()) {
+            builder.set(mapping.getTargetField(), lookupValue);
+          } else {
+            builder.set(mapping.getTargetField(), mapping.getDefaultValue());
+          }
+        }
+      } else {
+        // for the fields except source fields
+        builder.set(sourceField.getName(), input.get(sourceField.getName()));
+      }
+    }
+
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+    setLookup(context);
+  }
+
+  /**
+   * @param pipelineConfigurer
+   * @throws IllegalArgumentException when source field is other than String type
+   */
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    Schema outputSchema = null;
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    if (inputSchema != null) {
+      outputSchema = getOutputSchema(inputSchema);
+    }
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(outputSchema);
+
+  }
+
+  /**
+   * Object used to keep input mapping corresponding to each source field
+   */
+
+  public static class ValueMapping {
+
+    private Lookup<String> lookupTable;
+    private String targetField;
+    private String lookupTableName;
+    private String defaultValue;
+
+    public ValueMapping(){
+    }
+
+    public Lookup<String> getLookupTable() {
+      return lookupTable;
+    }
+
+    public void setLookupTable(Lookup<String> lookupTable) {
```

Comment:
see earlier comment.  It's better not to have setters when they are not needed.  All of these things are not expected to change, and can be final and set in the constructor.